File: Algorithm_141_Linked List Cycle.cpp
#include <iostream>
using namespace std;
struct ListNode
{
   int val;
   ListNode *next;
   ListNode(int x) : val(x), next(NULL) {}
};
bool hasCycle(ListNode *head)
{
	if (head==NULL || head->next==NULL) return false;
	ListNode *p1=head, *p2=head->next;
	while (p1!=NULL && p2!=NULL)
	{
		if (p1==p2) return true;
		p1=p1->next;
		if (p2->next!=NULL) p2=p2->next->next;
		else return false;
	}
	return false;
}

File: Algorithm_141_Linked List Cycle.java
// 思路：快慢指针。首先判断head是否为null，若为null则一定无环，返回false
// fast每次走2步，slow每次走一步，则fast相对于slow的速度为1
// 若原链表有环，则slow和fast最终一定会相遇，此时退出循环并返回true
// 若fast或fast.next为null，则一定无环，退出循环并返回false

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
}